File: backend/src/scoring/strategies/strategy-registry.ts
import { Injectable, Inject, Logger } from '@nestjs/common';
import {
  ScoringStrategy,
  StrategyConfig,
  SCORING_STRATEGY,
} from './scoring-strategy.interface';

@Injectable()
export class StrategyRegistry {
  private readonly logger = new Logger(StrategyRegistry.name);
  private readonly strategies: Map<string, ScoringStrategy> = new Map();
  private readonly configs: Map<string, StrategyConfig> = new Map();

  constructor(
    @Inject(SCORING_STRATEGY)
    strategies: ScoringStrategy[],
  ) {
    for (const strategy of strategies) {
      this.register(strategy);
    }
  }

  register(strategy: ScoringStrategy): void {
    this.strategies.set(strategy.name, strategy);
    this.configs.set(strategy.name, {
      name: strategy.name,
      enabled: true,
      weight: strategy.defaultWeight,
    });
    this.logger.log(`Registered strategy: ${strategy.name} (weight: ${strategy.defaultWeight})`);
  }

  getStrategy(name: string): ScoringStrategy | undefined {
    return this.strategies.get(name);
  }

  getEnabledStrategies(): Array<{ strategy: ScoringStrategy; config: StrategyConfig }> {
    const result: Array<{ strategy: ScoringStrategy; config: StrategyConfig }> = [];

    for (const [name, strategy] of this.strategies) {
      const config = this.configs.get(name);
      if (config?.enabled) {
        result.push({ strategy, config });
      }
    }

    return result;
  }

  getAllStrategies(): Array<{ strategy: ScoringStrategy; config: StrategyConfig }> {
    const result: Array<{ strategy: ScoringStrategy; config: StrategyConfig }> = [];

    for (const [name, strategy] of this.strategies) {
      const config = this.configs.get(name)!;
      result.push({ strategy, config });
    }

    return result;
  }

  updateConfig(name: string, updates: Partial<StrategyConfig>): void {
    const config = this.configs.get(name);
    if (!config) {
      throw new Error(`Strategy ${name} not found`);
    }

    this.configs.set(name, { ...config, ...updates });
    this.logger.log(`Updated strategy config: ${name}`, updates);
  }

  setWeight(name: string, weight: number): void {
    if (weight < 0 || weight > 1) {
      throw new Error('Weight must be between 0 and 1');
    }
    this.updateConfig(name, { weight });
  }

  enable(name: string): void {
    this.updateConfig(name, { enabled: true });
  }

  disable(name: string): void {
    this.updateConfig(name, { enabled: false });
  }

  getNormalizedWeights(): Map<string, number> {
    const enabled = this.getEnabledStrategies();
    const totalWeight = enabled.reduce((sum, { config }) => sum + config.weight, 0);

    const normalized = new Map<string, number>();
    for (const { config } of enabled) {
      normalized.set(config.name, totalWeight > 0 ? config.weight / totalWeight : 0);
    }

    return normalized;
  }
}

File: backend-java/src/main/java/com/peopleminer/scoring/StrategyRegistry.java
package com.peopleminer.scoring;

import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Component;

import java.util.*;
import java.util.stream.Collectors;

@Component
@Slf4j
public class StrategyRegistry {

    private final Map<String, ScoringStrategy> strategies = new LinkedHashMap<>();
    private final Map<String, StrategyConfig> configs = new HashMap<>();

    public StrategyRegistry(List<ScoringStrategy> scoringStrategies) {
        for (ScoringStrategy strategy : scoringStrategies) {
            register(strategy);
        }
    }

    public void register(ScoringStrategy strategy) {
        strategies.put(strategy.getName(), strategy);
        configs.put(strategy.getName(), StrategyConfig.builder()
            .name(strategy.getName())
            .enabled(true)
            .weight(strategy.getDefaultWeight())
            .build());
        log.info("Registered strategy: {} (weight: {})", strategy.getName(), strategy.getDefaultWeight());
    }

    public Optional<ScoringStrategy> getStrategy(String name) {
        return Optional.ofNullable(strategies.get(name));
    }

    public List<StrategyWithConfig> getEnabledStrategies() {
        return strategies.entrySet().stream()
            .filter(e -> Boolean.TRUE.equals(configs.get(e.getKey()).getEnabled()))
            .map(e -> StrategyWithConfig.builder()
                .strategy(e.getValue())
                .config(configs.get(e.getKey()))
                .build())
            .collect(Collectors.toList());
    }

    public List<StrategyWithConfig> getAllStrategies() {
        return strategies.entrySet().stream()
            .map(e -> StrategyWithConfig.builder()
                .strategy(e.getValue())
                .config(configs.get(e.getKey()))
                .build())
            .collect(Collectors.toList());
    }

    public void updateConfig(String name, StrategyConfig updates) {
        StrategyConfig config = configs.get(name);
        if (config == null) {
            throw new IllegalArgumentException("Strategy " + name + " not found");
        }

        if (updates.getWeight() != null) {
            config.setWeight(updates.getWeight());
        }
        if (updates.getEnabled() != null) {
            config.setEnabled(updates.getEnabled());
        }

        log.info("Updated strategy config: {} -> {}", name, config);
    }

    public void setWeight(String name, double weight) {
        if (weight < 0 || weight > 1) {
            throw new IllegalArgumentException("Weight must be between 0 and 1");
        }
        StrategyConfig config = configs.get(name);
        if (config != null) {
            config.setWeight(weight);
        }
    }

    public void enable(String name) {
        StrategyConfig config = configs.get(name);
        if (config != null) {
            config.setEnabled(true);
        }
    }

    public void disable(String name) {
        StrategyConfig config = configs.get(name);
        if (config != null) {
            config.setEnabled(false);
        }
    }

    public Map<String, Double> getNormalizedWeights() {
        List<StrategyWithConfig> enabled = getEnabledStrategies();
        double totalWeight = enabled.stream()
            .mapToDouble(swc -> swc.getConfig().getWeight())
            .sum();

        Map<String, Double> normalized = new HashMap<>();
        for (StrategyWithConfig swc : enabled) {
            double normalizedWeight = totalWeight > 0
                ? swc.getConfig().getWeight() / totalWeight
                : 0;
            normalized.put(swc.getConfig().getName(), normalizedWeight);
        }

        return normalized;
    }

    @Data
    @NoArgsConstructor
    @AllArgsConstructor
    @Builder
    public static class StrategyConfig {
        private String name;
        private Boolean enabled;
        private Double weight;
    }

    @Data
    @NoArgsConstructor
    @AllArgsConstructor
    @Builder
    public static class StrategyWithConfig {
        private ScoringStrategy strategy;
        private StrategyConfig config;
    }
}
